### PR TITLE
Send/Receive Headers via Http2 Frames

### PR DESCRIPTION
**Pull request by oldergod**
getting ready for #4267

### CODE REVIEW

File: okhttp-tests/src/test/java/okhttp3/internal/http2/Http2ConnectionTest.java [hunk 1]
```diff
@@ -539,6 +541,143 @@
     assertTrue(Arrays.equals("c3po".getBytes("UTF-8"), requestData.data));
   }
 
+  static final class RecordingHeadersListener implements Header.Listener {
```
Comment (swankjesse) [inline okhttp-tests/src/test/java/okhttp3/internal/http2/Http2ConnectionTest.java]: for follow-up we should move this into its own top-level file

Comment (swankjesse) [reply okhttp-tests/src/test/java/okhttp3/internal/http2/Http2ConnectionTest.java]: though by that time it’ll be a `List<Headers>` not a `List<Header>`!

**Comment by swankjesse** [comment]
```

[INFO] --- maven-checkstyle-plugin:2.17:checkstyle (default) @ okhttp ---
[INFO] Starting audit...
[ERROR] /home/travis/build/square/okhttp/okhttp/src/main/java/okhttp3/internal/http2/Http2Stream.java:391: Line is longer than 100 characters (found 103). [LineLength]
Audit done.
```

File: okhttp/src/main/java/okhttp3/internal/http2/Http2Stream.java [hunk 2]
```diff
@@ -343,64 +357,84 @@ synchronized void receiveRstStream(ErrorCode errorCode) {
...
+              headers = null;
+
+            } else if (readBuffer.size() > 0) {
+              // Prepare to read bytes. Start by moving them to the caller's buffer.
+              readBytesDelivered = readBuffer.read(sink, Math.min(byteCount, readBuffer.size()));
+              unacknowledgedBytesRead += readBytesDelivered;
+
+              if (errorCodeToDeliver == null
+                  && unacknowledgedBytesRead
+                  >= connection.okHttpSettings.getInitialWindowSize() / 2) {
```
Comment (oldergod) [inline okhttp/src/main/java/okhttp3/internal/http2/Http2Stream.java]: that's tough indention

**Pull request merged**